### DISCUSSION TITLE
Questions about implementation wasi-nn on hardware

### DISCUSSION THREAD

**Discussion by TAOFOR4** [WebAssembly/wasi-nn#68] [Q&A] [answered]
Hi, thx for this great project. I've been researching wasm, ml and iot. So I am wondering if it is possible to use wasi-nn on iot devices at the current stage. I saw there were some experiments on Intel processors, but I haven't seen any on IoT devices. Thank you!

**Comment by abrown** [comment]
wasi-nn is in no way limited to Intel processors. The key point of wasi-nn (and WASI in general) is to allow a WebAssembly program to access system APIs. How these APIs are implemented is up to the WebAssembly runtime and runtimes are available for several HW targets.

**Reply by TAOFOR4** [reply]
Thx for the quick response. Based on your explanation, I guess my question should be: Has any Webassembly runtime that supports wasi-nn implemented it on any IoT device like Raspberry Pi or so? Could you provide any insights or updates on the feasibility and current state of using wasi-nn on IoT platforms? Thank you for your time and efforts!

**Reply by abrown** [reply, marked as answer]
It certainly seems feasible to me but you may have to do some research here to understand this in more depth (I don't typically run much on a Raspberry Pi, e.g.). For completeness, you probably want to look into the various implementations of wasi-nn:
- [__WAMR__]: as you've discovered [here](https://github.com/bytecodealliance/wasm-micro-runtime/discussions/3221), it seems that their is previous work that has run wasi-nn programs on smaller devices
- [__Wasmtime__]: there is ARM support that should make this runnable on a Raspberry Pi; you would probably want to use the ONNX backend
- [__WasmEdge__]: you would have to ask around there; I'm not as familiar with that runtime

[__WAMR__]: https://github.com/bytecodealliance/wasm-micro-runtime
[__Wasmtime__]: https://github.com/bytecodealliance/wasmtime
[__WasmEdge__]: https://github.com/WasmEdge/WasmEdge

**Reply by TAOFOR4** [reply]
Got it! Now I have a better picture. Really appreciate your patient guidance. I will look into what you mentioned. Thank you!

**Comment by Saadtalsulami** [comment]
What specific hardware are you targeting (CPU, GPU, NPU, or AI accelerator)?
